c/compare.py

```diff
@@ -0,0 +1,174 @@
+
+import glob
+import yaml
+import os
+import shutil
+import subprocess
+from optparse import OptionParser
+
+
+def expand(s):
+    return sorted(glob.glob(s))
+
+
+def getArchitecture(cfgfile):
+    with open(cfgfile, "r") as fp:
+        for line in fp:
+            if line.startswith("Architecture"):
+                return line.split()[1]
+    print("Invalid configuration file", cfgFile)
+    exit(1)
+
+
+def getTasksFromSet(setFile):
+    with open(setFile, "r") as fp:
+        for line in fp:
+            if not line.startswith("#"):
+                for task in expand(line.strip()):
+                    yield task
+
+
+BLACKLIST = set()
+for f in ["floats-esbmc-regression/trunc_nondet_2.i", "*pthread*/*"]:
+    BLACKLIST.update(expand(f))
+
+# parse comand line options and set default values
+parser = OptionParser()
+parser.add_option("-k", "--keep-going", dest="KEEP_GOING", default=False, action="store_true")
+parser.add_option("-v", "--diff", action="store_true", default=False, dest="SHOW_DIFF")
+parser.add_option("--skip-large", action="store_true", default=False, dest="SKIP_LARGE")
+(options, args) = parser.parse_args()
+
+# if the user did not directly specify sets, we select all sets
+SETS=[]
+for arg in (args if args else ["*.set"]):
+    SETS.extend(expand(arg))
+
+# build goto-cc and goto-diff if not available, and then set PATH to find it
+if not shutil.which("goto-cc") or not shutil.which("goto-diff"):
+  if not os.path.exists("../cbmc.git/src/goto-cc/goto-cc"):
+    subprocess.call(["git", "clone" "--depth=1" "http://github.com/diffblue/cbmc.git" "../cbmc.git"])
+    subprocess.call(["make", "minisat2-download"], cwd="../cbmc.git/src")
+    subprocess.call(["make", "CXX=g++-5", "goto-diff.dir", "goto-cc.dir"], cwd="../cbmc.git/src")
+  cwd = os.getcwd()
+  os.environ['PATH'] = cwd + "/../cbmc.git/src/goto-cc:" + cwd + "/../cbmc.git/src/goto-diff:" + os.environ['PATH']
+
+EC=0
+
+# iterate over all sets
+for f in SETS:
+  if not os.path.exists(f):
+    print("Invalid set", f)
+    exit(1)
+
+  setf=os.path.basename(f)[:-4] # remove ending ".set"
+
+  # skip some sets, like LDV (too big) or Concurrency (pthread headers are very platform dependent)
+  if setf == "ConcurrencySafety-Main":
+    print("Skipping category", setf, "(platform-dependent types)")
+    continue
+  elif options.SKIP_LARGE and setf.startswith("Systems_DeviceDriversLinux64_ReachSafety"):
+    print("Skipping category", setf, "(only custom includes, no system headers, checking takes too much time)")
+    continue
+  elif setf == "Systems_OpenBSD_MemSafety":
+    print("Skipping category", setf, "(only custom includes, no system headers, complicated build process)")
+    continue
+  elif setf == "Systems_SQLite_MemSafety":
+    print("Skipping category", setf, "(complicated build process, requires patched version of cilly)")
+    continue
+
+  if not os.path.exists(setf + ".cfg"):
+    print("Skipping category", setf, "(no .cfg file present)")
```

Comment:
Shouldn't this be an error?

And we could move this code as well as the code that checks for a valid architecture into `getArchitecture`, then the main loop would be easier to read.